File: csharp/Stage1/Evaluator.cs
using System;
using System.Collections.Generic;

namespace MidLang.Stage1
{
    /// <summary>
    /// Evaluator (Interpreter)
    /// 
    /// Purpose: Executes the AST to produce program output.
    /// 
    /// How it works:
    /// 1. Traverses the AST nodes
    /// 2. Evaluates expressions (computes values)
    /// 3. Manages variable storage (symbol table)
    /// 4. Executes statements (assignments, prints)
    /// </summary>
    public class Evaluator
    {
        /// <summary>
        /// Symbol table: stores variable names and their values.
        /// This is like a dictionary: variable name → value
        /// </summary>
        private readonly Dictionary<string, int> _symbolTable = new Dictionary<string, int>();

        /// <summary>
        /// Evaluates a program by executing all its statements.
        /// </summary>
        public void Evaluate(ProgramNode program)
        {
            foreach (Statement statement in program.Statements)
            {
                EvaluateStatement(statement);
            }
        }

        /// <summary>
        /// Evaluates a statement (assignment or print).
        /// </summary>
        private void EvaluateStatement(Statement statement)
        {
            switch (statement)
            {
                case AssignmentStatement assign:
                    EvaluateAssignment(assign);
                    break;

                case PrintStatement print:
                    EvaluatePrint(print);
                    break;

                default:
                    throw new Exception($"Unknown statement type: {statement.GetType()}");
            }
        }

        /// <summary>
        /// Executes an assignment statement: stores the expression's value in the variable.
        /// </summary>
        private void EvaluateAssignment(AssignmentStatement assign)
        {
            int value = EvaluateExpression(assign.Expression);
            _symbolTable[assign.VariableName] = value;
        }

        /// <summary>
        /// Executes a print statement: evaluates the expression and outputs the result.
        /// </summary>
        private void EvaluatePrint(PrintStatement print)
        {
            int value = EvaluateExpression(print.Expression);
            Console.WriteLine(value);
        }

        /// <summary>
        /// Evaluates an expression and returns its integer value.
        /// </summary>
        private int EvaluateExpression(Expression expression)
        {
            return expression switch
            {
                IntegerLiteral lit => lit.Value,
                VariableReference varRef => EvaluateVariable(varRef),
                BinaryExpression binExpr => EvaluateBinaryExpression(binExpr),
                _ => throw new Exception($"Unknown expression type: {expression.GetType()}")
            };
        }

        /// <summary>
        /// Evaluates a variable reference: looks up the variable's value in the symbol table.
        /// </summary>
        private int EvaluateVariable(VariableReference varRef)
        {
            if (!_symbolTable.ContainsKey(varRef.Name))
            {
                throw new Exception($"Undefined variable: {varRef.Name}");
            }
            return _symbolTable[varRef.Name];
        }

        /// <summary>
        /// Evaluates a binary expression: evaluates left and right, then applies the operator.
        /// </summary>
        private int EvaluateBinaryExpression(BinaryExpression binExpr)
        {
            int left = EvaluateExpression(binExpr.Left);
            int right = EvaluateExpression(binExpr.Right);

            return binExpr.Operator switch
            {
                "+" => left + right,
                "-" => left - right,
                "*" => left * right,
                "/" => right == 0 ? throw new Exception("Division by zero") : left / right,
                _ => throw new Exception($"Unknown operator: {binExpr.Operator}")
            };
        }
    }
}

File: cpp/Stage1/Evaluator.cpp
#include "Evaluator.h"
#include <iostream>
#include <stdexcept>
#include <sstream>

void Evaluator::evaluate(ProgramNode* program) {
    for (auto& statement : program->statements) {
        evaluateStatement(statement.get());
    }
}

void Evaluator::evaluateStatement(Statement* statement) {
    if (auto* assign = dynamic_cast<AssignmentStatement*>(statement)) {
        evaluateAssignment(assign);
    } else if (auto* print = dynamic_cast<PrintStatement*>(statement)) {
        evaluatePrint(print);
    } else {
        throw std::runtime_error("Unknown statement type");
    }
}

void Evaluator::evaluateAssignment(AssignmentStatement* assign) {
    int value = evaluateExpression(assign->expression.get());
    symbolTable[assign->variableName] = value;
}

void Evaluator::evaluatePrint(PrintStatement* print) {
    int value = evaluateExpression(print->expression.get());
    std::cout << value << std::endl;
}

int Evaluator::evaluateExpression(Expression* expression) {
    if (auto* lit = dynamic_cast<IntegerLiteral*>(expression)) {
        return lit->value;
    } else if (auto* varRef = dynamic_cast<VariableReference*>(expression)) {
        return evaluateVariable(varRef);
    } else if (auto* binExpr = dynamic_cast<BinaryExpression*>(expression)) {
        return evaluateBinaryExpression(binExpr);
    } else {
        throw std::runtime_error("Unknown expression type");
    }
}

int Evaluator::evaluateVariable(VariableReference* varRef) {
    auto it = symbolTable.find(varRef->name);
    if (it == symbolTable.end()) {
        std::stringstream ss;
        ss << "Undefined variable: " << varRef->name;
        throw std::runtime_error(ss.str());
    }
    return it->second;
}

int Evaluator::evaluateBinaryExpression(BinaryExpression* binExpr) {
    int left = evaluateExpression(binExpr->left.get());
    int right = evaluateExpression(binExpr->right.get());

    if (binExpr->op == "+") {
        return left + right;
    } else if (binExpr->op == "-") {
        return left - right;
    } else if (binExpr->op == "*") {
        return left * right;
    } else if (binExpr->op == "/") {
        if (right == 0) {
            throw std::runtime_error("Division by zero");
        }
        return left / right;
    } else {
        std::stringstream ss;
        ss << "Unknown operator: " << binExpr->op;
        throw std::runtime_error(ss.str());
    }
}

